Fabricators will now produce items immediately if they are next in queue

boy2mantwicethefam opened the pull request:
You know that delay that makes building components be an unnecessary slog that nobody bothers to wait a few minutes for? This PR gets rid of it. Things will start being built immediately if they are next in queue, which means you can now print 50 components in about 10 seconds with a fully-upgraded Protolathe, which will make people start upgrading machineries.
Report any bugs that you can find, R&D code is a mess.
Current bugs:
If there are not enough materials for the next item in queue while the queue did not start with the item that cannot be produced, the item will be removed from the queue instead of being moved at the bottom.

:cl:
 * tweak: Drastically reduced the delay between objects being produced at fabricators, meaning that 50 stock parts will now take 10 seconds to build instead of approx. 2 minutes. Report any bugs you can find.

SonixApache commented:
EXTREME lathecreep / science buff but probably for the best all things considered
maybe slow down production speeds *slightly* to compensate

boy2mantwicethefam commented:
I could try adding a very small (less than a second) delay to not make it too spammable but that would depend if sleep or spawn work on decimals, which I will find out soon enough.
I want to see the current iteration in action, and if people abuse it to create a ton of items in a short timespan I will be fully for adding a small delay, which is as easy as adding a sleep or spawn as the first thing after checking if there's a queue length

SonixApache commented:
The no-delay is probably fine, but slowing down base production (x0.95?) would be a decent way to balance it probably.

boy2mantwicethefam commented:
That's too small of a change and will affect other items while stock parts will continue being produced very fast due to their very low material costs (the primary thing that affects production is material cost, the higher the material cost the more it takes to build, it's auto-generated), but I know what you mean. I will now try to add a delay of 0.3 seconds between every queue production, which is still an improvement over the byond tick

boy2mantwicethefam commented:
During testing spawn had a weird behavior that made at least two items disappear from the queue, so sleep was used instead.
The delay has been reduced to 0.2 seconds instead, which comes off as 5 components per second at 50 components per 10 seconds, which tickles my pseudo-OCD.

DrSnips commented:
@SonixApache What's even the balance concern here? This realistically only affects stuff with small costs and made in bulk like stock parts and med kits. Oh no they can be useful to the station in a timely manner.

SonixApache commented:
something something speedrunning the job so you can go autism
whatever

codeungabunga commented:
Delays for sake of delays make no sense.

boy2mantwicethefam commented:
It's ready to be merged, because alternatives did not work
Current bug:
- Item disappears from queue list if it can't be built

Bugs from alternatives:
Added a check for !build_part that places the thing at the end of the queue
- All prior items built placed in queue (IE: 5 capacitors and one unavailable ultra laser, 5 capacitors placed at the end of queue)

Replaced !build_part(D) with a new !build_part(I), I being the same as D but not called for before that
- Item converts next item from queue list into itself or something like that

Placed a spawn(0) and then !build_part(I)
- Queue no longer works

It's ready to be merged, with a note attached.

ShiftyRail commented:
I hope you did test it

boy2mantwicethefam commented:
If I didn't test it I couldn't give these accurate bug reports for alternatives
